### ISSUE TITLE
A token with VBL/light moving doesn't expose hard fog for other tokens

### ISSUE THREAD

**Issue by Merudo** [RPTools/maptool#934]
**Is your feature request related to a problem? Please describe.**
If a token with a light source or VBL moves, new area of the map can become visible but the hard fog isn't cleared.

The classic example would be opening a door, in which case the room should become visible. However the hard fog isn't exposed when this happens - the room remains black.

It would be nice if the hard fog would automatically become exposed by the movement of other tokens.

**Describe the solution you'd like**
If a token with VBL or light source moves, a new check for exposure happens.

For performances reasons, we might want to only expose the hard fog for PCs tokens.

**Additional context**
I'm not sure if this feature is too performance intensive to implement.

**Comment by Azhrei** [comment]
> For performances reasons, we might want to only expose the hard fog for PCs tokens.

I don't think we can do that (NPCs might be owned by players and if they Have Vision, the players should be able to see the exposed areas).  However, we only need to check tokens with Has Vision turned on...

> **Additional context**
> I'm not sure if this feature is too performance intensive to implement.

Even if it is, it's The Right Thing To Do, so we should probably have an implementation, even if there's an option that defaults to off.

**Comment by Phergus** [comment]
Note that if a player owned token is moved with the right server settings then hard fog is cleared.

Auto-exposure of new visible areas when a GM moves a token that has VBL and/or a light source would probably require a similar server setting.

**Comment by Merudo** [comment]
> Note that if a player owned token is moved with the right server settings then hard fog is cleared.

With individualized FOW, it is only cleared for that one token though. The other tokens owned by the player don't get the exposure.

**Comment by Phergus** [comment]
True but that's more of a bug with IV/IFoW than a new feature.  Automatic exposure because a non-player controlled light source or something with VBL moved is definitely not expected currently and would be a new feature.

**Comment by aliasmask** [comment]
I think this is a bug. https://forums.rptools.net/viewtopic.php?f=6&t=28522&p=275218#p275216

Before, if a light source appeared or a PC token moved with a light source exposing an area another PC can see with light source, it would expose the FoW. This is not happening now. Tested with 1.5.10.

**Comment by JamzTheMan** [comment]
When you say "before", what version worked as expected, so we can compare?

**Comment by JamzTheMan** [comment]
Also, would have to check but IIRC iFoW is tracked by token GUID so there probably is a loop test to combine and match by owner?
Maybe instead we could use the client ID to cache the results?

**Comment by aliasmask** [comment]
I'll have to hop in to the way back machine. Before 1.4.8

**Comment by aliasmask** [comment]
So, it's been awhile. I think ever since the option to 1.3.b76 had, but 1.3.b84 doesn't. So, I'm thinking ever since the reveal on move server start option was added. The years are just a blur.

The reason I think it's a bug is if you're in a dark room and can't see and someone with a light source comes in, you still can't see unless you select your token (if you can) and hit ctrl-i. I just think in the past I've just moved player tokens if they couldn't see.

**Comment by Phergus** [comment]
Here's how it worked in previous versions:
```
1.3.b87 - Player token light source exposed for all players on movement.
          NPC token light source DID NOT EXPOSE for players on movement.
1.3.b89 - Player token light source DID NOT EXPOSE for all players on movement.
1.3.b91 - Player token light source DID NOT EXPOSE for all players on movement.
1.4.0.5 - Player token light source DID NOT EXPOSE for all players on movement.
1.5.11  - Player token light source DID NOT EXPOSE for all players on movement.
          NPC token light source DID NOT EXPOSE for players on movement.
```
MT 1.3.b87 has IV and IFoW.

**Comment by Azhrei** [comment]
The 1.3b87 seems like that's the best approach, but I would want areas exposed if the NPC were player-owned.  (I will sometimes give light sources to an NPC and place them off the side of the map, only moving them onto the map when I'm ready.  I wouldn't want that movement to auto-expose as GM.)

Does that seem right?

**Comment by JamzTheMan** [comment]
Seems silly that a PC couldn't see a NPC with a torch. So I'm looking down a hallway and a guard walks across the hallway with a torch, the PC never saw that intersection?

Smells wrong. 

But at least it's been that way for a while. When people report a bug I'm always like, "oh gosh, what did we break? Oh, it's been like that for 3+ years. Ok, not a bug. That's a feature request."

**Comment by Phergus** [comment]
I should point out that in 1.3.b87, it exposed not just the area that other Player tokens would be able to see when the light passed through but also areas that it couldn't.   This was with IV and IFoW enabled.

**Comment by aliasmask** [comment]
FYI, this is my work around. Set a token property to 
`*mouseoverEvent:[H: exposeFOW()]` 
and lighting will update when mousing over your token. Token does have to be selected unless you put currentToken() as parameter. There could be reasons to have it one way an not the other like preventing players from exposing area around them when going to new map since they usually can't select their token easily until it's exposed.

**Comment by Merudo** [comment]
@aliasmask wouldn't it be more efficient to use `onTokenMove` instead?

**Comment by jfrazierjr** [comment]
Im' not sure what @aliasmask use case it, but ontokenmove does not(by intent) happen for NPC tokens and certain other criteria I forget

**Comment by aliasmask** [comment]
I use my own onTokenMove and already do this, but it doesn't expose for the other players. By mousing over your own token ensures the FoW is updated. One example is when the GM removes some VBL say to open a door. The map doesn't refresh the FoW. Players are always fiddling with their tokens and instead of having to move it or hitting ctrl-i, they can just mouseover their token to update. It's not perfect, but it's convenient.

**Comment by aliasmask** [comment]
I'm also only talking about what PC see. NPC don't typically use their vision. I do turn on NPC vision briefly to establish targeting, but that doesn't reveal FoW. Only the NPC light source is useful for the PCs.
